### PR TITLE
added print template endpoint

### PR DESCRIPTION
**Pull request by LukeLoze1**
# Motivation and Context
New endpoint to get and create print templates

# What has changed
New endpoint and some test helper classes

# How to test?
use create to add some 

e.g. post:  http://localhost:7777/api/printtemplates
body
  {
      "packCode": "abc1",
       "template": ["hello", "world"],
      "printSupplier": "PrintSupplier2"
  }

Then use get to see if you can retrieve them

Get: http://localhost:7777/api/printtemplates

# Links
https://trello.com/c/OWzk3XOT/2629-rops-ui-second-feature-create-print-template-api-5

# Screenshots (if appropriate):

### CODE REVIEW

File: src/main/java/uk/gov/ons/ssdc/responseoperations/endpoint/PrintTemplateEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    printTemplate.setPackCode(printTemplateDto.getPackCode());
+    printTemplate.setTemplate(printTemplateDto.getTemplate());
+    printTemplate.setPrintSupplier(printTemplateDto.getPrintSupplier());
+
+    printTemplateRepository.saveAndFlush(printTemplate);
+
+    return new ResponseEntity(HttpStatus.CREATED);
+  }
+
+  private PrintTemplateDto mapPirntTemplates(PrintTemplate printTemplate) {
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/responseoperations/endpoint/PrintTemplateEndpoint.java:58]:
```suggestion
  private PrintTemplateDto mapPrintTemplates(PrintTemplate printTemplate) {
```

File: src/main/java/uk/gov/ons/ssdc/responseoperations/endpoint/PrintTemplateEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  public ResponseEntity createPrintTemplate(
+      @RequestBody PrintTemplateDto printTemplateDto,
+      @Value("#{request.getAttribute('userEmail')}") String userEmail) {
+    userIdentity.checkGlobalUserPermission(
+        userEmail, UserGroupAuthorisedActivityType.CREATE_PRINT_TEMPLATE);
+
+    PrintTemplate printTemplate = new PrintTemplate();
+    printTemplate.setPackCode(printTemplateDto.getPackCode());
+    printTemplate.setTemplate(printTemplateDto.getTemplate());
+    printTemplate.setPrintSupplier(printTemplateDto.getPrintSupplier());
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/responseoperations/endpoint/PrintTemplateEndpoint.java:51]: the print supplier is not being validated. It needs to be checked against the print supplier config. See how Support Tool's API does it here: https://github.com/ONSdigital/ssdc-rm-support-tool/blob/3c8c65dbee7be38789452b82c5f9a4512c0a4211/src/main/java/uk/gov/ons/ssdc/supporttool/endpoint/PrintSuppliers.java#L44

File: src/main/java/uk/gov/ons/ssdc/responseoperations/endpoint/PrintTemplateEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  @GetMapping
+  public List<PrintTemplateDto> getPrintTemplates(
+      @Value("#{request.getAttribute('userEmail')}") String userEmail) {
+    userIdentity.checkGlobalUserPermission(
+        userEmail, UserGroupAuthorisedActivityType.LIST_PRINT_TEMPLATES);
+
+    List<PrintTemplate> printTemplates = printTemplateRepository.findAll();
+    return printTemplates.stream().map(this::mapPirntTemplates).collect(Collectors.toList());
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/responseoperations/endpoint/PrintTemplateEndpoint.java:38]:
```suggestion
    return printTemplates.stream().map(this::mapPrintTemplates).collect(Collectors.toList());
```

**Review by NickGrantONS — CHANGES_REQUESTED** [review]
Needs to validate print supplier against configured suppliers

File: src/main/java/uk/gov/ons/ssdc/responseoperations/endpoint/PrintTemplateEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  @PostMapping
+  public ResponseEntity createPrintTemplate(
+      @RequestBody PrintTemplateDto printTemplateDto,
+      @Value("#{request.getAttribute('userEmail')}") String userEmail) {
+    userIdentity.checkGlobalUserPermission(
+        userEmail, UserGroupAuthorisedActivityType.CREATE_PRINT_TEMPLATE);
+
+    if (!checkPrintSupplierValid(printTemplateDto.getPrintSupplier())) {
+      return new ResponseEntity(HttpStatus.BAD_REQUEST);
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/responseoperations/endpoint/PrintTemplateEndpoint.java:57]: why doesn't the `checkPrintSupplierValid` method just throw the exception... it lives inside the Endpoint class anywawy

**Review by NickGrantONS — CHANGES_REQUESTED** [review]
Sorry... another couple of validations needed: pack code can't be blank/null. Also, obviously there must be a minimum of one non-blank/null item in the template array... indeed every item in the array must not be blank/null, and there has to be one item for it to be a meaningful/useful template

**Review by NickGrantONS — APPROVED** [review]

**Review by liamtoozer — APPROVED** [review]
LGTM 👍

**Pull request merged**